### PR TITLE
Deprecate QuantumEngineSampler

### PR DESCRIPTION
**Pull request by dstrain115**
- Deprecates this class in favor of ProcessorSampler
- Moves get_engine_sampler into engine with the other utility functions
- Change engine.get_sampler to only support a single processor and have
  it point to the processor's sampler.

Fixes: #5371

### CODE REVIEW

**Comment by dstrain115** [comment]
@mrwojtek FYI: QuantumEngineSampler is being deprecated in favor of ProcessorSampler.  Let me know if you see any problems with the changes to the calibration workflow code.

File: cirq-google/cirq_google/engine/engine.py [hunk 1]
```diff
@@ -877,3 +884,30 @@ def get_engine_calibration(
...
+            ``Engine.list_processors``).
+        gate_set_name: One of ['sqrt_iswap', 'sycamore'].
+            See `cirq_google.NAMED_GATESETS`.
+        project_id: Optional explicit Google Cloud project id. Otherwise,
+            this defaults to the environment variable GOOGLE_CLOUD_PROJECT.
+            By using an environment variable, you can avoid hard-coding
+            personal project IDs in shared code.
+
+    Returns:
+        A `QuantumEngineSampler` instance.
```
Comment (maffoo) [inline cirq-google/cirq_google/engine/engine.py:906]: nit: I'd suggest removing this, since it doesn't add anything beyond the return type annotation. If you want to leave it, can just say it returns a sampler.

File: cirq-google/cirq_google/engine/engine.py [hunk 2]
```diff
@@ -782,7 +781,7 @@ def get_processor(self, processor_id: str) -> engine_processor.EngineProcessor:
     @util.deprecated_gate_set_parameter
     def sampler(
         self, processor_id: Union[str, List[str]], gate_set: Optional[Serializer] = None
-    ) -> engine_sampler.QuantumEngineSampler:
+    ) -> cirq.Sampler:
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/engine/engine.py]: Yeah, it's probably better for this to be more specific.

File: cirq-google/cirq_google/engine/engine.py [hunk 3]
```diff
@@ -801,21 +800,29 @@ def sampler(
     @util.deprecated_gate_set_parameter
     def get_sampler(
         self, processor_id: Union[str, List[str]], gate_set: Optional[Serializer] = None
-    ) -> engine_sampler.QuantumEngineSampler:
+    ) -> cirq.Sampler:
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/engine/engine.py]: Done.

File: cirq-google/cirq_google/engine/engine.py [hunk 4]
```diff
@@ -877,3 +884,30 @@ def get_engine_calibration(
     May return None if no calibration metrics exist for the device.
     """
     return get_engine(project_id).get_processor(processor_id).get_current_calibration()
+
+
+def get_engine_sampler(processor_id: str, project_id: Optional[str] = None) -> cirq.Sampler:
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/engine/engine.py]: Done.

File: cirq-google/cirq_google/engine/engine.py [hunk 5]
```diff
@@ -877,3 +884,30 @@ def get_engine_calibration(
...
+    """Get an EngineSampler assuming some sensible defaults.
+
+    This uses the environment variable GOOGLE_CLOUD_PROJECT for the Engine
+    project_id, unless set explicitly.
+
+    Args:
+        processor_id: Engine processor ID (from Cloud console or
+            ``Engine.list_processors``).
+        gate_set_name: One of ['sqrt_iswap', 'sycamore'].
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/engine/engine.py]: Done.

File: cirq-google/cirq_google/engine/engine.py (hunk 1 above)
Comment (dstrain115) [reply cirq-google/cirq_google/engine/engine.py]: Done.

File: cirq-google/cirq_google/engine/engine.py [hunk 6]
```diff
@@ -877,3 +884,30 @@ def get_engine_calibration(
...
+        project_id: Optional explicit Google Cloud project id. Otherwise,
+            this defaults to the environment variable GOOGLE_CLOUD_PROJECT.
+            By using an environment variable, you can avoid hard-coding
+            personal project IDs in shared code.
+
+    Returns:
+        A `QuantumEngineSampler` instance.
+
+    Raises:
+         ValueError: If the supplied gate set is not a supported gate set name.
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/engine/engine.py]: Done.

File: cirq-google/cirq_google/calibration/workflow.py [hunk 7]
```diff
@@ -826,9 +826,9 @@ def run_calibrations(
 
     if isinstance(sampler, Engine):
         engine: Optional[Engine] = sampler
-    elif isinstance(sampler, QuantumEngineSampler):
-        engine = sampler.engine
-        (processor_id,) = sampler._processor_ids
+    elif isinstance(sampler, ProcessorSampler):
+        processor_id = getattr(sampler.processor, 'processor_id', None)
+        engine = sampler.processor.engine() if processor_id is not None else None
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/calibration/workflow.py]: Yeah, there were some typing shenanigans that made this necessary.  I refactored so that the function below takes a processor instead of an engine, which removes the need for this.

**Review by maffoo — APPROVED** [review]

File: cirq-google/cirq_google/engine/engine.py [hunk 8]
```diff
@@ -877,3 +884,26 @@ def get_engine_calibration(
     May return None if no calibration metrics exist for the device.
     """
     return get_engine(project_id).get_processor(processor_id).get_current_calibration()
+
+
+def get_engine_sampler(
```
Comment (mpharrigan) [inline cirq-google/cirq_google/engine/engine.py:889]: Why was this moved? Just for better organizational sense?

**Comment by mpharrigan** [comment]
nice; when I was doing #5361 I was worried the calibration dispatching would be a mess; but you've done it very nicely
